### DISCUSSION TITLE
How to build in parallel several targets *sharing common code* ?

### DISCUSSION THREAD

**Discussion by Fabrizio-Caruso** [cc65/cc65#2445] [Q&A] [answered]
CC65 seems to make it harder or impossible to build the same game for different targets or different games when these share some common source files.
I don't have the same problem with gcc and Z88DK, which both make parallel builds work with no issue.
I suppose the problem is that CC65 generates some temporary .o files on the same directory as the .c, which it also deletes. This creates race conditions on such files:
```
ld65: Error: Input file 'cross_lib/display/display_macros.o' not found
ld65: Error: Input file 'cross_lib/display/display_macros.o' not found
ld65: Error: Input file 'cross_lib/display/display_macros.o' not found
Fatal error: Cannot open input file 'cross_lib/sound/cc65/vic20/vic20_sounds.s': No such file or directory
ld65: Error: Input file 'cross_lib/display/display_macros.o' not found
ld65: Error: Input file 'cross_lib/display/display_macros.o' not found
make: *** [Makefile_common:628: vic20_exp_16k] Error 1
make: *** [Makefile_common:628: vic20_exp_16k] Error 1
make: *** [Makefile_common:628: vic20_exp_16k] Error 1
make: *** [Makefile_common:628: vic20_exp_16k] Error 1
make: *** [Makefile_common:628: vic20_exp_16k] Error 1
make: *** [Makefile_common:628: vic20_exp_16k] Error 1
ld65: Error: Input file 'cross_lib/display/display_macros.o' not found
make: *** [games/shoot/makefiles/Makefile.override:9: vic20_exp_16k] Error 1
```
Is there a way to tell CC65 to avoid this? Either by putting such .o in memory or in a separate temporary folder per target or by using some sort of locks or any other parallel-friendly mechanism?

If this is not possible yet, this would be a nice feature to add in the future. 

I asked the Z88DK devs how they do it. They say that when I run their compiler with several source files in one line, no .o is generated.

**Comment by mrdudz** [comment]
The proper fix for this would be changing cc65 to use proper unique temp files.... obviously that doesn't really help you now though :)

Not sure how i would solve this

**Comment by Fabrizio-Caruso** [comment]
This is what Z88DK devs did to make their cross-compiler work well with parallel builds:
[@](https://github.com/z88dk/z88dk/issues/2527)

So they put all .o in /tmp and they must be doing it with some unique names.

**Comment by rofl0r** [comment]
that's a pretty standard thing to do, not rocket science. POSIX has several functions to deal with this, tmpnam, tmpfile, mkstemp, mkdtemp, and probably even others.

**Reply by Fabrizio-Caruso** [reply]
It would be quite useful to anyone who is using cc65 to build for several targets at the same time and probably anyone who want to make parallel builds.
I hope this capability is added to cc65 one day.

**Reply by rofl0r** [reply]
i've written a wrapper for a windows C compiler and had to implement my own "create unique tmp filename" function because the standard library's tmpnam() was too basic. maybe studying the code can give you insight and even allow you to implement the feature for cc65.
https://github.com/rofl0r/pellescc/blob/master/pellescc.c#L147
note that even that code which is pretty unlikely to generate the same name for more than one concurrent process sometimes creates a collision, i've experienced a build error due to this in about 1 out of 40 times building a project with about 50 C files. probably should have used the process id (on unix via getpid()) too. that said, such workarounds are only needed for compilers with a very limited standard library, on mainstream compilers/libc like in linux or BSD you can usually depend on tmpnam() doing the right thing on its own.

**Comment by Fabrizio-Caruso** [comment]
@mrdudz could this discussion be marked as feature request?
It is a quite general feature that would bring benefits to everyone.

**Comment by mrdudz** [comment, marked as answer]
fixed by #2660
